pennylane/wires.py

```diff
@@ -30,30 +30,36 @@ def _process(wires):
         # if input is already a Wires object, just return its wire tuple
         return wires.labels
 
-    elif isinstance(wires, (Number, str)):
+    if isinstance(wires, (Number, str)):
         # interpret as a single wire
         return (wires,)
 
-    elif hasattr(wires, "shape") and wires.shape == tuple():
+    if hasattr(wires, "shape") and wires.shape == tuple():
         # Scalar NumPy array
         return (wires.item(),)
 
-    elif isinstance(wires, Iterable) and all(isinstance(w, Wires) for w in wires):
-        # if the elements are themselves Wires objects, merge them to a new one
-        return tuple(w for wires_ in wires for w in wires_.tolist())
-
-    elif isinstance(wires, Iterable) and all(
-        isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires
-    ):
-        # if the elements are strings or numbers, turn iterable into tuple
-        return tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
+    while isinstance(wires, Iterable):
```

Comment:
How come this is a while loop? From my reading of the below, it seems like, at most, only a single loop will ever execute:

* Either an if statement is activated and the `merged` list is returned, or

* None of the if statements are activated, and the while loop is exited via the `break`.

Unless I'm missing additional logic, could we replace this with an if statement?
```suggestion
    if isinstance(wires, Iterable):
```

Reply:
One can replace the `while` loop by an `if` loop at the cost of duplicating some code. 
Firstly, one could replace the `break` by a copy of the `WireError` raised at the end of `_process` if no condition applied.
Secondly one could only ever enter the `if isinstance(wires, Iterable)` loop if one of the secondary conditions also applies, resulting in duplication of the check for these secondary conditions (whether all elements are a Wires instance / a scalar).
Thirdly, one could go back to the previous structure of the `if` clauses and duplicate the check for `isinstance(wires, Iterable)`. 
Effectively, the `while` loop is used as an `if` loop in which one can use `break`, which is kind of hacky, I must admit.

Reply:
Got it! Thanks @dwierichs.

It's quite a neat solution to avoid the duplicating code. In this case, however, I'm tempted to say that perhaps code readability is more important than avoiding duplicating the exception? On my first reading of the code, I assumed that there was a loop occurring, and when I realized there wasn't, I thought the `while` was a mistake!

Perhaps another approach could be to factor out the `iterable` validation to a separate function:

```python
def _process_iterable(wires):
    if all(isinstance(w, Wires) for w in wires):
        # if the elements are themselves Wires objects, merge them to a new one
        merged = tuple(w for wires_ in wires for w in wires_.tolist())
    elif all(isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires):
        # if the elements are strings or numbers, turn iterable into tuple
        merged = tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
    else:
        return None

    if len(set(merged)) != len(merged):
        raise WireError("Wires must be unique; got {}.".format(merged))

    return merged
```

and then, within `_process`,

```python
if isinstance(wires, Iterable):
    wires = _process_iterable(wires)
    if wires is not None:
        return wires

raise WireError(...)
```

Reply:
Okay, I agree it is not as clear as the other solutions. I actually like the "`else return None` --> `if not None return merged`" approach. But maybe we can leave it in `_process`, like so:
```
    if all(isinstance(w, Wires) for w in wires):
        merged = tuple(w for wires_ in wires for w in wires_.tolist())
    elif all(isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires):
        merged = tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
    else:
        merged = None
    if merged is not None:
        if len(set(merged)) != len(merged):
            raise WireError("Wires must be unique; got {}.".format(merged))
        return merged
raise WireError(...)
```

Reply:
Just to understand, is there a speed difference in this rewrite too? 

If so, I like the last version you posted most @dwierichs.

Reply:
I agree, the last version you have @dwierichs is the clearest.

It is also probably faster than the function version I wrote above, since function calls in Python always add a non-negligible overhead.

Reply:
> It is also probably faster than the function version I wrote above, since function calls in Python always add a non-negligible overhead.

Yes, that should be true. 

> Just to understand, is there a speed difference in this rewrite too?

The ordered decisions I made were: 
- remove check for uniqueness from __init__
- therefore have to add it to the cases werer `isinstance(wires, Iterable)` (of which there were two previously)
- therefore unify the two iterable scenarios and write down the test for uniqueness only once.
So this change is part of the speed-up, yes. I hope this answers the question :)

Reply:
> If so, I like the last version you posted most 

I updated `_process` to this version now.